### PR TITLE
Introduce as-shot to later D65 workflow (improved highlights)

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
Replaces #15461 and #15505

**Please note that temperature module version is increased in the PR so keep backups while testing**

The current modern chroma correction workflow assumes "camera reference point" (nomally D65) has the best available data for temperature coeffs as preliminary for all modules in the pixelpipe before colorin. There we change to working profile and continue work for best visual quality until the "color calibration" module.

There is one drawback with this approach. Some modules in the pipe would like to have "perfect white balance" correction - the rgb channels all have the same value for any greytone.

Examples:
- Some hightlights reconstruction algos take the other channels or surrounding data into account and modify data "towards white".
- raw chromatic aberration correction also iterates from channel differences. Good "white is white" coeffs help significantly here.
- Some demosaicers also have slightly improved output on pixelpeeping level.
- denoise profiled

In short - how does it work?
```
was: data(rgb) -> temperature(*D65)     -> ModA -> ModB ...                   ->colorin
now: data(rgb) -> temperature(*as_shot) -> ModA -> ModB ... -> *(D65/as_shot) ->colorin
```

This version has more anti-chroma-casting implemented for opposed and segmentation based, generally better than we had before.

### CODE REVIEW

**Comment by TurboGit** [comment]
I'll let other test too... but on my side I see no difference using D65 or late-D65 mode, using snapshot and comparing on multiple images (zoom 100%, 200% & 400%) it seems that every pixel is identical on the two modes.

**Comment by jenshannoschwalm** [comment]
Take the image you suggested, reduce exposure so that you don't have anything bright any more and maybe even disable filimc/sigmoid.

Look or possibly zoom in to the bright all-blown area at the top left.

Now toggle between the mode 4 / 5 (the new one) and either a) observe the vectorscope/histo b) use a picker there or c) use the snapshot !

There will be differences **only** if modules work not in a strictly linear way like highlights (the others are not always as "easy" to spot as here), @rawfiner gave another example with denoising, i also could provide images with CA where you will have slightly better reconstruction).

**Comment by jenshannoschwalm** [comment]
@TurboGit not sure if we would want the new mode to be the new default...

**Comment by IainMF** [comment]
> 1. @rawfiner certainly worth more testing if you find the time
> 2. @wpferguson for you too
> 3. @garagecoder and @IainMF it might be interesting for you too and i would appreciate your testing on highlights. I did a lot of tests and i couldn't yet find an image that's worse with this code than what we had before. (Almost all images make blown outs more towards white than before ...)

This sounds interesting. Unfortunately, I'd need someone to do a windows build for me.

One thing I thought of was, what happens when the 'as shot' white balance is wrong. For example, if you used the 'daylight' setting on your camera when the lighting was 'tungsten'.

**Comment by jenshannoschwalm** [comment]
> One thing I thought of was, what happens when the 'as shot' white balance is wrong.

Yes - that's the major problem not making this the 'default' :-)

**Comment by TurboGit** [comment]
@jenshannoschwalm : Indeed lowering the exposure I see now that the late D65 gives a slightly better result (less reddish).

About the question for making the new mode the default, my proposal would be:
- Make the new mode the default
- For the new edit, have a single button for the new mode
- For old edits, keep both buttons
- When new mode is selected for an old edit, keep only the new button when entering darkroom again.

Of course this is only if new mode is always better. What do you think about this?

But... Then I'd like to do that at a safe time and not now, so I'm proposing to wait for after the 4.6 release.

What about this plan?

**Comment by jenshannoschwalm** [comment]
Fully agreed. 4.8

**Comment by garagecoder** [comment]
Checking by eye, there's a general improvement for modules such as highlight recovery when in the new mode. The idea seems sound, but perhaps difficult to test output for correctness.

**Comment by wpferguson** [comment]
Tested the latest changes.  Works fine for me...

**Comment by jenshannoschwalm** [comment]
@wpferguson IIRC you build on windows too. Could you provide a windows version including this for @IainMF ?

**Comment by rawfiner** [comment]
If I understand correctly, the main disadvantage of this mode is when the "as shot" white balance is wrong. Actually, would it be possible to synchronise the white balance module with color calibration settings instead of using the "as shot" white balance? That way, it would garantee good results in all cases as the user will be able to fix wrong white balance settings.

**Comment by jenshannoschwalm** [comment]
> Actually, would it be possible to synchronise the white balance module with color calibration settings instead of using the "as shot" white balance?

What do you mean here "exactly" ?

**Comment by rawfiner** [comment]
Yes, some sort of feedback.
Color calibration would decide the correct white balance, the white balance module would apply a similar one, that would be transformed to D65 before colorin, as in your PR.
Note that I have no idea how doable this is though...

**Comment by wpferguson** [comment]
@IainMF here's a link to a windows version with the PR, https://drive.google.com/file/d/1fNRfxbjaN351ZrvjDJPf66EbHARjehZ0/view?usp=sharing

**Comment by IainMF** [comment]
> @IainMF here's a link to a windows version with the PR, https://drive.google.com/file/d/1fNRfxbjaN351ZrvjDJPf66EbHARjehZ0/view?usp=sharing

Thanks!

**Comment by TurboGit** [comment]
@jenshannoschwalm : This needs a rebase an conflict resolution.

I've a bit lost track of the work here, what about having this new mode the default as usually better (I have proposed this here https://github.com/darktable-org/darktable/pull/15602#issuecomment-1804235407).

Or do you prefer having this in now and rework the default later (or not if not actually a good idea)?

**Comment by jenshannoschwalm** [comment]
@TurboGit i will certainly rebase pretty soon and will do a lot more testing. About being the default, maybe i will be sure after that testing, otherwise we can decide later after more people - using other cameras and camera settings - had their say on this.

**Comment by TurboGit** [comment]
@jenshannoschwalm : Sounds like a plan!

**Comment by jenshannoschwalm** [comment]
> Sounds like a plan!

:-) Also there is more urgent bugfix work to do for 4.6.1, besides the RGBE loader i also spotted problems with some exr and tiff files :-(

**Comment by jenshannoschwalm** [comment]
@TurboGit @rawfiner @wpferguson 

Would appreciate your re-review after a lot of testing and more fixes.

1. The new mode has been set as default now, a) couldn't find any issues b) subtly better in almost all cases c) coudn't yet find a single image where it's worse
2. BTW the latest commit is something we might also want for 4.6.1 - or a working variant of it - as it seems to be a genuine bug

**Comment by AxelG-DE** [comment]
Gentlemen, many things here are above my head. I have tested a bit, as it is merged now.

One thing drew my attention. Pictures which have an empty/discarded history, the color calibration now is set to _adaptation = none (bypass)_ by default.

I donno this is a "has to be", just as a user I am medium happy about this, as most of the new pictuers I open, I first have to click the reset button on color cal....

**Comment by jenshannoschwalm** [comment]
@AxelG-DE was just asking you to write an issue :-)

**Comment by AxelG-DE** [comment]
I will

**Comment by RawConvert** [comment]
Does this work take into account using the "Unity" white balance method for expose-to-the-right shooting?  That is, you have your camera white balance set to (1,1,1).  So selecting "as shot" in DT gives a green image.  I'm thinking this will throw those modules which are supposed to like As Shot coefficients.

**Comment by jenshannoschwalm** [comment]
Sincerely, i dont understand. Some modules like demosaic, Highlights, denoise and RAW cacorrect simply need good  color coeffs for White Balance.

**Comment by RawConvert** [comment]
It's about judging clipping in the camera, typically for ETTR, it's not precise because the camera shows clipping in relation to the jpeg, not the Raw.  So someone had this idea to remove the white balance from the equation by using a custom white balance which happens to be (1,1,1), got by shooting a white-ish sheet of paper and totally overexposing.  So in DT, the as shot coeffs are (1,1,1).  But you don't use those obviously, you choose Camera Reference, or maybe set the temp that looks best.  The new white balance option is using As Shot, so I thought this might cause problems for those of us doing Uni White Balance.

Thinking about it more generally, for the majority not doing Uni W.B, if you're shooting Raw, do you care much about white balance?  I'd imagine many don't set it shot by shot, knowing it will be dealt with in DT.  The new option seems to be the default, so people will be using a "random" white balance for some of the modules.  That doesn't seem right.  It would be using a piece of data for the processing which the user never intended to be used, surely?

There are a number of articles about Uni WB, here's one  http://www.guillermoluijk.com/tutorial/uniwb/index_en.htm
The bit about overexposing a white sheet is at the end.

**Comment by jenshannoschwalm** [comment]
I know the dt code for all modules handling raw data quite well, most of them expect white balance to be as good as possible already. So if your preset is good, that's fine. (this means any Grey area should have equal rgb data after the temperature module.) if not, all those modules will work but with inferior quality. That was the rationale behind the pr, we make the best guess from available data.

Otherwise we use data found in exif, the new mode is slightly better than what we had if camera did correct wb, after colorin module the data are exactly as before this pr by default.

**Comment by RawConvert** [comment]
Hanno,  I'm struggling to see how your reply answers my points, in particular that the new option will use As Shot coefficients which may be at least imperfect and possibly way out compared with the scene when the photo was taken. Of course I don't know how many DT users don't worry about the camera WB, but do you have a feel for this?  It might be lots.

You say "most of them [the DT modules] expect white balance to be as good as possible already" and if not there will be "inferior quality".  So As Shot is not a guarantee of best quality, which is surely what we want DT to be capable of.  It seems to me that users need to set the best temp and tint in the WB module from the start, if they want best quality.  How about this as a different solution -

- communicate clearly to users the issue about needing best WB for certain modules for optimum quality
- continue to default WB according to Preferences settings
- no new option in WB dropdown!
- if user doesn't want to set temp and tint for rather small improvements in image quality (I understand it is marginal?) then user continues as previously, e.g. with Camera Reference
- if user wants best quality, user adjusts temp and tint
- there is a new tickbox in the module saying "use Camera Reference/D65 later in the processing" which is ticked by default.

I've tried a couple of raws and am a bit confused.  Both have As Shot as (1,1,1) i.e. Uni White Balance method. I can see no difference between "camera reference" and "as shot to reference".

Also when I choose "as shot to reference", I get a "white balance missing" red message in WB and color calibration.

**Comment by RawConvert** [comment]
Above is with Darktable-4.7.0+761.gb4eaf01d54-x86_64.AppImage

**Comment by wpferguson** [comment]
Some modules (profiled denoise, highlight reconstruction, etc) use the as shot white balance coefficients to achieve a more accurate result than when the camera reference white balance coefficients are used.  Color calibration is still used to white balance the image.

**Comment by RawConvert** [comment]
@wpferguson , I've gathered that's the case re. your first sentence, but what's the point you're making pls?  My point is that the As Shot coeffs may be inaccurate so be careful about using them, and educate the user about this issue...

**Comment by wpferguson** [comment]
> but what's the point you're making pls?

You can turn it off

**Comment by RawConvert** [comment]
Do you mean:  It will default to the new option when the raw is opened, then if the user is aware, they might go into WB and change it to camera reference.

**Comment by jenshannoschwalm** [comment]
I think you miss a point. Until now we used "camera reference white balance coefficients" as default in modern color workflow for **all** modules. (Of course you can change that in temperature module as you wish).

For the vast majority of images the "as shot" parameters are slightly better a) for raw data processing and b) profiled denoise so those are now using them, colorin has been modified and changes to "camera reference white balance coefficients" further on in the pipe. This is again default only for modern color workflow. So what this PR does:
Slightly improves quality for the majority of images, a few images might be slightly worse.

You can of course use any of the other ways or even use the "non modern" workflow.
 

> I've tried a couple of raws and am a bit confused. Both have As Shot as (1,1,1) i.e. Uni White Balance method. I can see no difference between "camera reference" and "as shot to reference".

So what? You should see no big difference. Watch out for demosaicer pixel problems, try profiled denoise, try to handle highlights .. you will spot them. As said, very subtle but reality.

**Comment by RawConvert** [comment]
Hanno, you say "For the vast majority of images the "as shot" parameters are slightly better [for some modules]"  Pls can you explain this in more detail.  I mean, how do you justify that?  Are you assuming the vast majority will have been shot with an accurate white balance setting in the camera?  To me this is a big unknown.  For some time now we users have been told the modern workflow using camera reference and color calibration is the way to go, and I haven't doubted that, but I think many users will have realised that the white balance setting in their camera is pretty/totally irrelevant for this Raw workflow.  So they just leave it on daylight or auto or something, and often it may be quite inappropriate for the specific modules in question.  If lots of users are doing this, then I think it's wrong to implement this new option as the default for scene-referred process.

**Comment by RawConvert** [comment]
"It is the way to go with modern work flow and camera reference."
	That's quite a general statement, I have no comment!

"This is also not guaranteed to be safe BTW."
	Same as above

"We now assume that - we also do so in the old work flow and we did before in dt - that as-shot has likely more correct raw data, only in that stage of the pipe."
	Suppose we're using the current DT, 4.6, and scene-referred workflow therefore using Camera Reference in WB.  Are the As Shot coeffs used anywhere as part of calculating the final export image?
	
"Later we use the same color maths. OK, it's an assumption, yet I tested on a vast number of images, PlaRaw, dpreview, public available test images..."
	Did you consider publicising this in the Pixls forum and inviting users to try it / test it?  Wouldn't that have been a useful exercise?

"I couldn't yet find a file failing."
	By failing do you mean causing a crash, or producing a sub-optimal image, or what?

"Would be happy if you could provide such a bad file."
	I'll try to find some time to try some raws, especially high iso and denoise module.  I'm not trying to prove your new feature is no good, I'm just challenging what it's based on and how it gets implemented.

"Do you have problems with your work flow?"
	No, I'm very happy with it apart from the fact I could do with a GPU.  If your feature is implemented as I understand it, I'm wondering what will happen if I adjust temp and tint in WB module to get a good look, in order for those "sensitive" modules to work best.  Then use color calibration (even if it doesn't need to do much).  Is there any disadvantage to working this way, e.g. in other parts of the pipe, since the coeffs will no longer be the same as camera reference?  Can you make it so this works well?!

**Comment by jenshannoschwalm** [comment]
> By failing do you mean causing a crash, or producing a sub-optimal image, or what?

By failing i meant: anything worse qualitywise or with a significant color cast different from old default.

 

> Suppose we're using the current DT, 4.6, and scene-referred workflow therefore using Camera Reference in WB. Are the As Shot coeffs used anywhere as part of calculating the final export image?

I think i already said that: Nope



> I'm wondering what will happen if I adjust temp and tint in WB module to get a good look

It will fallback to "user changed" as before. When developing that part i experimented with the temp&tint parameters to only change the data in the pre-colorin part of the pixelpipe. Can indeed give even better results ... Left that out as users will be irritated that "image look" will change not in a way visible without pixel-peeping.

 

> I'll try to find some time to try some raws, especially high iso and denoise module

Cool. if you find "something" you can share maybe do it here or email me for privacy.

**Comment by todd-prior** [comment]
> It will fallback to "user changed" as before. When developing that part i experimented with the temp&tint parameters to only change the data in the pre-colorin part of the pixelpipe. Can indeed give even better results ... Left that out as users will be irritated that "image look" will change not in a way visible without pixel-peeping.

I am just a little unclear about this statement ... so in the case where as shot might be off but you do have a good neutral target and you select that in the wb module is that "user" value used for the modules being calculated prior to using the D65 reference values or are you saying its not used. I am just curious as that could be a "better" wb for those modules that benefit from good wb data to perform better. Maybe that is what you are saying...

**Comment by jenshannoschwalm** [comment]
Maybe we are "overcomplicating" ...

1. The whitebalance module generally works as before. It just has a new button.
2. The first 4 buttons work exactly the same as before. Also changing **any** of the sliders work exactly as before.
3. Old edits won't change at all.

So what is new?
If you select "as shot to reference" manually - or by default in modern temperature workflow - the pixelpipe uses as-shot data for modules **before** "colorin".  **In**  "colorin" we correct in two steps, 1) to camera reference D65 and 2) to selected colorspace. **After** colorin it's all the same.

The summary would be: As "as-shot" coeffs are mostly better suited for "pre colorin" modules than "camera reference" we use them in that part of the pixelpipe. If you find an image with inferior quality just select the old default (or change manually)

Personally i couldn't find an image with inferior quality with the new mode compared to "camera reference". We could have that if "as-shot" is completely wrong. To check watch out (pixelpeeping!) for single pixels being wrong in LMMSE, raw cacorrection (pixelpeeping again), for colors off after segmentation based highlights or for denoised profiled (again pixelpeeping). But mostly these modules behave better.

**Comment by RawConvert** [comment]
Hanno,   please bear with this a little further, nearly there I think!
Maybe complicating but colour science and DT are complicated!
So we have two groups of modules - several that work best with an accurate temp/tint - let's call these the "sensitive" ones; and the rest which frequently operate with Camera Reference / D65, this having been defined in the WB module.  Consider a user who wants the best possible image quality.  Suppose they adjust temp/tint accurately, thinking this is their best plan.  The WB module will now show "user modified", not "as shot to reference".  Then they adjust the rest of the modules for the image.

The sensitive modules will now be optimal.  But what about the rest?  I remember Aurelien Pierre and people talking about needing D65 for certain modules to be optimal, and Camera Reference became the scene-referred default.  So in my scenario above, are the not-sensitive modules working at their best image quality?

If not, can they be made to, so the user has the best possible image quality available in DT?

**Comment by RawConvert** [comment]
Hanno - I was just trying the latest nightly appimage - 4.7 806.  The tint slider is going to 1.598 rather than 1 if you double-click it.

**Comment by jenshannoschwalm** [comment]
Will be back on Monday...

**Comment by todd-prior** [comment]
Thanks for the expanded explanation....


Sent from 
Outlook for Android

**Comment by jenshannoschwalm** [comment]
@RawConvert 


> The WB module will now show "user modified", not "as shot to reference".

Yes. This means all modules later in the pipe will behave exactly as they did without this pr.



> The sensitive modules will now be optimal. But what about the rest? I remember Aurelien Pierre and people talking about needing D65 for certain modules to be optimal, and Camera Reference became the scene-referred default. So in my scenario above, are the not-sensitive modules working at their best image quality?
> 
> If not, can they be made to, so the user has the best possible image quality available in DT?

This depends on "color calibration" module.

I definitely wanted to avoid more "controls" in the temperature module, 1) that is already a very tricky beast code-wise and 2) user questions would likely explode :-)

**Comment by RawConvert** [comment]
"This depends on "color calibration" module."
Please can you expand on this, I don't understand.

**Comment by jenshannoschwalm** [comment]
> Please can you expand on this, I don't understand.

I am awfully sorry but that has been the whole story of the modern color workflow, having D65 or Camera Reference for "color calibration" and all foloowing modules. That's why we normally have two modules for "color-adaptation". Please see manual or tutorials or earlier comments on this :-) But be sure, that is still the case with the new default introduced here.

**Comment by RawConvert** [comment]
So regarding use of D65 where needed, it's fine for "as shot to reference".  But 7 posts above I'm asking about "user modified".  My understanding is that with user modified, D65 will NOT be used for the module(s) that need it in order to work best.  This is my main concern.  I think what you are proposing is not the best solution.

As shot and user modified should be the same workflow in the sense that where D65 coefficients are needed, **both** should use them, and where temp/tint coeffs are needed, **both** should use those.

Above you refer to an explosion of user questions if you add a new control.  I think you might well get lots of questions anyway.  For example -
1. Why make this change if the differences to raw chromatic abs, highlights etc are very small.  I don't think you've posted any convincing pics or parts of pics that really make the case.
2. You're ensuring D65 coeffs are used later in the pipe for the new option, which might be using quite inappropriate temp/tint values earlier in the pipe; but you're not using D65 later in the pipe for user modified, where the user has taken the trouble to get the temp/tint coeffs right.  This is illogical and prevents the user getting the best image quality.
3. Users might ask why camera reference doesn't remain the default, because they have lots of raws taken with little attention to the camera white balance, because it has not been part of the scene-referred workflow until now.
4. Surely point 2, the main issue I think, can be resolved by keeping the four current dropdown options and adding a checkbox which is saying "use D65 coeffs where appropriate". You say this is tricky, it's only a tickbox for goodness sake!

**Comment by todd-prior** [comment]
Could this not be a way to simplify it. so there are only three controls As shot user modified and wb to reference. User option would be a selection based on picker or sliders. The first two could be used in legacy and there is only one reference mode. Old edits could just be supported but going forward the later conversion to d65 is considered superior and the only option needed and uses whichever wb mode of the two is selected?

**Comment by RawConvert** [comment]
I haven't yet digested todd-prior's comment above, but I've just thought of a further point re. my previous comment.
Perhaps there's no need for a checkbox anyway.
Is it quite simple? - just apply D65 coeffs later in the pipe in all cases where the user has selected scene-referred in Preferences.
Camera reference - yes!
As shot - yes
User modified - yes
From image area - yes

No need for a new option in the dropdown.

**Comment by jenshannoschwalm** [comment]
> My understanding is that with user modified, D65 will NOT be used for the module(s) that need it in order to work best. This is my main concern. I think what you are proposing is not the best solution.

Right. BUT that has not been changed by this pr at all and was not intended.

> As shot and user modified should be the same workflow in the sense that where D65 coefficients are needed, **both** should use them, and where temp/tint coeffs are needed, **both** should use those.

Simply don't get your point here. Again this has not changed at all in this pr ...




> 2\. You're ensuring D65 coeffs are used later in the pipe for the new option, which might be using quite inappropriate temp/tint values earlier in the pipe; but you're not using D65 later in the pipe for user modified, where the user has taken the trouble to get the temp/tint coeffs right.  This is illogical and prevents the user getting the best image quality.

Again not getting your point. This again has not changed at all.



> 3\. Users might ask why camera reference doesn't remain the default, because they have lots of raws taken with little attention to the camera white balance, because it has not been part of the scene-referred workflow until now.

Ok. The answer would be: For modern workflow we always use D65 camera reference either by using "Camera reference" or "as shot to reference" being the default. I guess some confusion might be because of the "naming" and tooltip here, will think about that.

**Comment by jenshannoschwalm** [comment]
> 1. I don't think you've posted any convincing pics or parts of pics that really make the case

You might look at #15505, two examples about denoising.

**Comment by todd-prior** [comment]
> . The answer would be: For modern workflow we always use D65 camera reference either by using "Camera reference" or "as shot to reference" being the default. I guess some confusion might be because of the "naming" and tooltip here, will think about that.

Can you comment on why we would have the two reference modes.... if we introduce the second one to use as shot wb for the desired modules as this is better for the pipeline why would it not replace the full reference mode using only D65 values... Or is there a use case still for the older way.... And I could see why as shot would be the default to use in a hybrid model like that introduced but if the As-shot is baddly determined should there not be a whay to correct wb and still use the hybrid approach.... And I may just be completely naive and missing something.... so not a critisism just trying to clarify the rationale in my head. Because then it would seem we have a method in place that is potentially better for the calculations in the pipeline but can only be used if you use as shot wb which might be terrible for a given shot and then you would have to resort to the old reference method or modified wb in legacy mode..... my apology for being thick

**Comment by RawConvert** [comment]
Hanno, ok yes, I see the examples, that's good.

Re. your other comments, I think we need a bit of a "reset".  You're saying certain things haven't changed in this PR, and that's true, but I think the PR itself is not right.  I expect you've put a fair bit of time and effort into this, but please try to step back from the PR and take a fresh look at the issue.  For a start, please see what I say below about this issue and see if you can view it in a different way...

Several years ago scene-referred processing was introduced and as part of that, the use of D65 coefficients for certain modules.  Many users have been very happy with the results produced.  Since then, last Autumn?, it seems people have found Camera Reference/D65 doesn't give the best results for certain modules like denoise, highlight reconstruction, and these need accurate white balanced coefficients for best results. DT strives to give great image quality.  So what is to be done?  For users doing scene-referred processing, they need two things -
- be able to use a good white balance in the modules that need it
- be able to use D65 coeffs in the later modules which need D65.

They could achieve the first by using one of the 3 options in the WB module apart from Camera Reference.  They can use Camera Reference if they don't want to do any white balancing at this stage, and accept a possible slight hit on image quality.

They can achieve the second, D65, without thinking, by having it baked into DT.  That is, if they are doing scene-referred, then D65 is used later in the pipe where needed, period.

I can think of only one scenario not covered by the above.  This is if the user wants to use the coeffs from the WB module throughout the pipe.  I don't think this qualifies as DT scene-referred processing, but perhaps the devs might want to support it.  In which case, perhaps selecting "none" for the auto-applied pixel workflow defaults could be the trigger for not using D65 at all.

**Comment by jenshannoschwalm** [comment]
> Can you comment on why we would have the two reference modes....

@todd-prior you gave the answer mostly yourself. If your image white balance is good-to-perfect the new mode is better, if it's not, it's better to use the "camera reference" mode.

Another short excursion about "why do we need good white balance" for all pre-demosaic modules (also profiled denoise,  not going into that) I would call it the "grey should be gray" rule. Meaning: in that part of the pipeline, a grey photo pixel should ideally have the same signal for each RGB channel. (Sidenote again: camera reference D65 is often slightly worse, a point AP probably oversaw when introducing the modern color workflow)

A few examples from code (memory) i know and understand well:
1. In the new HLR modes like inpaint opposed we "guess" clipped signals from unclipped and increase the clipped data towards-white. You see? The better that works, the less magenta we have. The old modes where pretty rude, clipping for example clipped all data so every rgb signal was the same :-) I didn't evaluate laplacian here so can't comment.
2. In raw ca correction the situation is not as obvious but we also need a good idea about "what is white". Do you remember the old days before i re-implemented it to use **all** image data? Without that the chance for a good correction was pretty bad.
3. Some demosaicers are more prone to this problem.
a) VNG sometimes introduces some color shift - mostly to brownish - you can see this on beach photos (not lava!)
b) LMMSE at one stage looks for the median of 9 values, this might go wrong with bad white coeffs. To spot this look at high-noise-low-signal images. Or as this is more stable against moire effects, look at feathers in bird images.
c) RCD and Amaze are not affected much, Amaze has one poart in the algo that looks suspicious but i couldn't find "anything"
4. As these effects are mostly subtle it would be good to test in "toggle high quality processing" mode. The internal scaling especially with the lanc modes is often misleading because of over/under-shoots


Andrew,
> Re. your other comments, I think we need a bit of a "reset"

:-) :-)
I agree in many parts here. It would be good to have a better "temperature" module UI.
1. I confess i truely hate the module code. Some parts are absolutely weird and it's a mess we can't do presets, styles on this. It might love a complete overhaul and your UI ideas might be pretty good & simple. But it works, warning about pipeline errors are correct now, ... :-)
2. At least i have been able to "fix" the "rawspace" part of color coeffs in the pipe. That was the important part for me, still evaluating more ideas about highlights correction. Fingers crossed ... 
3. We all have limited resources and my amount of work on dt is not hat bad :-) Not asking for donations at all, if you wanna spend some money, it would all be to support our ukrainian friends. 

So - i hope i finally have given enough information and inputs for further thinking. It would now be your part to go into coding :-) Certainly i would help out if you wanna dig into temperature hell!

**Comment by todd-prior** [comment]
Thanks for he extensive feedback... I fear I am still missing something. I am not debating the need for the use of wb rather I suspect it is beneficial as you have identified. I am merely saying that as-shot might not provide that so then you have to revert to the reference values when in fact you might still be able to benefit from this approach and you might be able to provide that good or desired wb on your own say from a gray card or whatever and then use the late introduction of reference values...

Perhaps I could frame it more like this and still I could be missing a key point...

Throwing caution to the wind I would simplify the wb module...to me the eyedropper and user modified are redundant and likely something that needs to be kept track of in code. I would move the picker next to the temp tint sliders more like it is found in most all other modules of DT and if you want a spot or area wb you go there as you do in other modules. This would leave 3 icons/controls as shot, user-modified, and whatever the new mode would be called but its basically the wb to late reference (ie your new mode).

If you move the sliders manually or do a wb from the picker the result is the same the sliders move to new values and the wb is not as shot anymore...why do we need both of these distinct modes to describe the same result... 

So from that you are down to 3 icons in WB module. It would start with two activated as-shot and wb to reference.

If the user so chooses to make manual changes and or use the picker to refine wb the pencil or user modified icon would show active instead of as shot and that wb would be used instead of as shot. So you use as-shot or modified wb for the wb sensitive modules and pass whichever one is active to the scene reference mode... 

Do we need the old reference only method if we are making the clear statement that other modules benefit from using wb ....

Likely this reveals how I think things work but I guess it might also reveal things I don't understand.... 

No need to reply...just my comments for what they are worth... I can work with whatever implementation occurs... Thanks for looking in to improving things.... All your work on the background code running DT should be heaped with praise.....

**Comment by RawConvert** [comment]
Hanno,  my coding days are long gone!  So I will have to leave this with yourself and the dev team.  Thanks for taking the time to reply and comment.

Todd, I suppose the WB GUI could be re-worked or the module re-written, but that isn't the issue here, surely, and it sounds like a big job.  The approach I've been pushing requires zero changes to the user interface layout or the controls within it.  It needs the later modules to know when to use D65 or the WB coeffs, but I think that is needed for Hanno's solution as well.

**Comment by jenshannoschwalm** [comment]
Thanks for all the discussion! We all have to make sure "camera presets" are used by many users and a preset taken should be clearly visible in the module UI.

**Comment by jenshannoschwalm** [comment]
@elstoc maybe this would be something for the manual :-)
________________________

The new default "as shot to reference" is a subtle modification of the "camera reference" mode.

Assuming in the majority of raw images the exif data for "as-shot" white balance are a better match for correct rgb coefficients than those found as "camera reference" we use those coefficients in the pixelpipe until we set "color input profile".
This results in slightly improved quality of modules like demosaic, denoised profile or highlights reconstruction but keeps all modules after "color input profile" working as in modern "camera reference" mode.

If for some reason the "as-shot" data are bad, use either the old "camera reference" default or use the other settings.
________________________________
What do you think?
